File: app/src/main/java/com/dmko/pairwisecomparison/ui/screens/comparisons/ComparisonsFragment.java
package com.dmko.pairwisecomparison.ui.screens.comparisons;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.dmko.pairwisecomparison.R;
import com.dmko.pairwisecomparison.data.entities.Comparison;
import com.dmko.pairwisecomparison.ui.base.mvp.impl.BaseFragment;
import com.dmko.pairwisecomparison.ui.screens.addeditcomparison.AddEditComparisonDialog;
import com.dmko.pairwisecomparison.ui.screens.comparison.ComparisonActivity;
import com.dmko.pairwisecomparison.ui.screens.comparisons.recyclerview.ComparisonsAdapter;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class ComparisonsFragment extends BaseFragment implements ComparisonsContract.View {

    @BindView(R.id.recycler_comparisons) RecyclerView recyclerComparisons;
    @BindView(R.id.progress_loading) ProgressBar progressLoading;
    @BindView(R.id.text_empty_title) TextView textEmptyTitle;
    @BindView(R.id.text_empty_description) TextView textEmptyDescription;

    @Inject ComparisonsContract.Presenter presenter;
    @Inject ComparisonsAdapter adapter;

    private Unbinder unbinder;

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getControllerComponent().inject(this);
        presenter.attachView(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_comparisons, container, false);
        unbinder = ButterKnife.bind(this, view);
        setupRecyclerView();
        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        presenter.start();
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.stop();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public void setComparisons(List<Comparison> comparisons) {
        adapter.setComparisons(comparisons);
        recyclerComparisons.setVisibility(View.VISIBLE);
        textEmptyTitle.setVisibility(View.GONE);
        textEmptyDescription.setVisibility(View.GONE);
    }

    @Override
    public void setEmptyComparisons() {
        recyclerComparisons.setVisibility(View.GONE);
        textEmptyTitle.setVisibility(View.VISIBLE);
        textEmptyDescription.setVisibility(View.VISIBLE);
    }

    @Override
    public void showLoading(boolean isLoading) {
        progressLoading.setVisibility(isLoading ? View.VISIBLE : View.GONE);
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public void showAddEditComparisonDialog(String comparisonId) {
        DialogFragment dialog = AddEditComparisonDialog.newInstance(comparisonId);
        showDialog(dialog);
    }

    @Override
    public void openComparison(Comparison comparison) {
        Intent intent = ComparisonActivity.getIntent(getContext(), comparison.getId(), comparison.getName());
        startActivity(intent);
    }

    public void onFabAddClicked() {
        presenter.addComparisonSelected();
    }

    @SuppressWarnings("ConstantConditions")
    private void setupRecyclerView() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        recyclerComparisons.setLayoutManager(layoutManager);
        recyclerComparisons.setAdapter(adapter);

        DividerItemDecoration itemDecoration = new DividerItemDecoration(getContext(), layoutManager.getOrientation());
        recyclerComparisons.addItemDecoration(itemDecoration);
    }
}

File: app/src/main/java/com/dmko/pairwisecomparison/ui/screens/comparisons/ComparisonsFragment.kt
package com.dmko.pairwisecomparison.ui.screens.comparisons

import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.dmko.pairwisecomparison.R
import com.dmko.pairwisecomparison.data.entities.Comparison
import com.dmko.pairwisecomparison.ui.base.mvp.impl.BaseFragment
import com.dmko.pairwisecomparison.ui.screens.addeditcomparison.AddEditComparisonDialog
import com.dmko.pairwisecomparison.ui.screens.comparison.ComparisonActivity
import com.dmko.pairwisecomparison.ui.screens.comparisons.recyclerview.ComparisonsAdapter
import com.dmko.pairwisecomparison.ui.screens.copycomparison.CopyComparisonDialog
import kotlinx.android.synthetic.main.fragment_comparisons.*
import javax.inject.Inject

class ComparisonsFragment : BaseFragment(), ComparisonsContract.View {

    @Inject lateinit var presenter: ComparisonsContract.Presenter
    @Inject lateinit var adapter: ComparisonsAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        controllerComponent.inject(this)
        presenter.attachView(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        val view = inflater.inflate(R.layout.fragment_comparisons, container, false)
        setupRecyclerView()
        return view
    }

    override fun onStart() {
        super.onStart()
        presenter.start()
    }

    override fun onStop() {
        super.onStop()
        presenter.stop()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.detachView()
    }

    override fun setComparisons(comparisons: List<Comparison>) {
        adapter.setComparisons(comparisons)
        recycler_comparisons.visibility = View.VISIBLE
        text_empty_title.visibility = View.GONE
        text_empty_description.visibility = View.GONE
    }

    override fun setEmptyComparisons() {
        recycler_comparisons.visibility = View.GONE
        text_empty_title.visibility = View.VISIBLE
        text_empty_description.visibility = View.VISIBLE
    }

    override fun showLoading(isLoading: Boolean) {
        progress_loading.visibility = if (isLoading) View.VISIBLE else View.GONE
    }

    override fun showAddEditComparisonDialog(comparisonId: String?) {
        val dialog = AddEditComparisonDialog.newInstance(comparisonId)
        showDialog(dialog)
    }

    override fun openComparison(comparison: Comparison) {
        val intent = ComparisonActivity.getIntent(context, comparison.id, comparison.name)
        startActivity(intent)
    }

    override fun openCopyComparisonDialog(comparisonId: String) {
        val dialog = CopyComparisonDialog.newInstance(comparisonId)
        showDialog(dialog)
    }

    fun onFabAddClicked() {
        presenter.addComparisonSelected()
    }

    private fun setupRecyclerView() {
        val layoutManager = LinearLayoutManager(context)
        recycler_comparisons.layoutManager = layoutManager
        recycler_comparisons.adapter = adapter

        val itemDecoration = DividerItemDecoration(context, layoutManager.orientation)
        recycler_comparisons.addItemDecoration(itemDecoration)
    }
}